Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.cpp

```diff
@@ -152,24 +153,61 @@ void RemoteDevice::createSampler(const WebGPU::SamplerDescriptor& descriptor, We
     m_objectHeap.addObject(identifier, remoteSampler);
 }
 
+static void populateConvertedDescriptor(auto mediaIdentifier, auto& convertedDescriptor, auto& performWithMediaPlayerOnMainThread)
+{
+#if ENABLE(VIDEO) && PLATFORM(COCOA)
+    using MediaPlayerOrVideoFrameResult = std::variant<WebCore::MediaPlayerIdentifier, WebKit::RemoteVideoFrameReference>;
+    MediaPlayerOrVideoFrameResult result = WTF::switchOn(mediaIdentifier, [] (PAL::WebGPU::HTMLVideoElementIdentifier i) -> MediaPlayerOrVideoFrameResult {
+        return WebCore::MediaPlayerIdentifier(i.identifier);
+    }, [] (PAL::WebGPU::WebCodecsVideoFrameIdentifier i) -> MediaPlayerOrVideoFrameResult {
+        return RemoteVideoFrameReference(RemoteVideoFrameIdentifier(i.identifier.first), i.identifier.second);
+    });
+
+    performWithMediaPlayerOnMainThread(result, [&convertedDescriptor](RefPtr<WebCore::VideoFrame> videoFrame) mutable {
+        convertedDescriptor->pixelBuffer = videoFrame ? videoFrame->pixelBuffer() : nullptr;
+    });
+#else
+    UNUSED_PARAM(mediaIdentifier);
+    UNUSED_PARAM(convertedDescriptor);
+    UNUSED_PARAM(performWithMediaPlayerOnMainThread);
+#endif
+}
+
+#if ENABLE(VIDEO) && PLATFORM(COCOA)
+void RemoteDevice::setSharedVideoFrameSemaphore(IPC::Semaphore&& semaphore)
+{
+    m_sharedVideoFrameReader.setSemaphore(WTFMove(semaphore));
```

Comment:
Wow, really? We need semaphores to use external textures? Why?

Reply:
This is how `SharedVideoFrameReader / Writer` is implemented. It is similar to what is done in `RemoteGraphicsContextGL`

Note we only encounter this path for local memory external textures. For external textures decoded in the GPU Process, we will not encounter this path.